lib/vminsertapi/client.go

```diff
@@ -0,0 +1,154 @@
+package vminsertapi
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/bytesutil"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/consts"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/encoding"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/encoding/zstd"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/handshake"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/storage"
+)
+
+// SendToConn sends given buf over provided bc to the server
+func SendRPCRequestToConn(bc *handshake.BufferedConn, rpcName string, buf []byte) error {
+	var err error
+	sizeBuf := sizeBufPool.Get()
+	defer sizeBufPool.Put(sizeBuf)
+
+	timeoutSeconds := len(buf) / 3e5
+	if timeoutSeconds < 60 {
+		timeoutSeconds = 60
+	}
+	timeout := time.Duration(timeoutSeconds) * time.Second
+	deadline := time.Now().Add(timeout)
+	if err := bc.SetWriteDeadline(deadline); err != nil {
+		return fmt.Errorf("cannot set write deadline to %s: %w", deadline, err)
+	}
+
+	rpcNameBytes := bytesutil.ToUnsafeBytes(rpcName)
+	sizeBuf.B, err = sendData(sizeBuf.B, bc, rpcNameBytes)
+	if err != nil {
+		return fmt.Errorf("cannot write rpcName %q: %w", rpcName, err)
+	}
+
+	zb := zbPool.Get()
+	defer zbPool.Put(zb)
+
+	zb.B = zstd.CompressLevel(zb.B[:0], buf, 1)
+	sizeBuf.B, err = sendData(sizeBuf.B, bc, zb.B)
+	if err != nil {
+		return fmt.Errorf("cannot write buf with size %d: %w", len(zb.B), err)
+	}
+	if err := bc.Flush(); err != nil {
+		return fmt.Errorf("cannot flush data with size %d: %w", len(zb.B), err)
+	}
+
+	// Wait for error message from vmstorage.
+	// This guarantees that the message has been fully received by vmstorage.
+	deadline = time.Now().Add(timeout)
+	if err := bc.SetReadDeadline(deadline); err != nil {
+		return fmt.Errorf("cannot set read deadline for reading `ack` to vmstorage: %w", err)
+	}
+	sizeBuf.B, err = readBytes(sizeBuf.B[:0], bc, maxErrorMessageSize)
+	if err != nil {
+		return fmt.Errorf("cannot read error message: %w", err)
+	}
+	if len(sizeBuf.B) > 0 {
+		if string(sizeBuf.B) == storage.ErrReadOnly.Error() {
+			return storage.ErrReadOnly
+		}
+		return errors.New((string(sizeBuf.B)))
+	}
+
+	return nil
+}
+
+// SendToConn sends given buf over provided bc to the server
```

Comment:
Add a comment that this is a legacy method and should not be used except for fallback

Reply:
I think, it's still could be useful for some cases.